[proposal] Add `auth` as a capability with support for OAuth 2.0 and secrets. (#101)
modelcontextprotocol/specification: docs/specification/auth/credential.md

---

> title: Credential Authentication

Reviewer: IMO this term is too ambiguous, since don't _all_ forms of "client giving details to the server" (including OAuth) involve some kind of credentials?

What about "token authentication" or "bearer authentication" or something like that? Neither is _quite_ expansive enough (e.g., to include use cases like providing a raw password), but that might be OK.

Author: I have no strong feelings about naming; my take is that OAuth is a fancy wrapper on top of this. Let's see if anyone is an advocate for a particular name and I can go rename the file + references to "credential".

type: docs
weight: 52
---

{{< callout type="info" >}}
**Protocol Revision**: {{< param protocolRevision >}}
{{< /callout >}}

The Model Context Protocol (MCP) supports credential-based authentication for scenarios requiring API keys or similar secrets.

## Capabilities

Clients supporting credential authentication **MUST** declare it during initialization:

```json
{
  "capabilities": {
    "auth": {
      "credential": true
    }
  }
}
```

Servers supporting credentials **MUST** include their capabilities:

```json
{
  "capabilities": {
    "auth": {
      "credential": {
        "list": true
      }
    }
  }
}
```

## Protocol Messages

### Credential Requirements

> Servers can list required credentials using the credential/list capability:

Reviewer: Is this meant to be s/capability/method/? If so, can we provide an example of that request?

```json
{
  "jsonrpc": "2.0",
  "id": 1,
  "result": {

>     "credential": [

Author: Leaving unresolved until naming is settled

      {
        "name": "X-API-KEY",
        "description": "An API key must be provided to call this tool."
      },
      {
        "name": "X-MISC-PASSWORD",
        "description": "A password must be provided to list this resource"
      }
    ]
  }
}
```

### Providing Credentials

> Clients provide credentials through headers during initialization:

Reviewer: `headers` is a bit confusing for non-HTTP transports, IMO. I think this should go into a `auth.credentials` field (or whatever we pick as the new name).

```json
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "initialize",
  "params": {
    "capabilities": {
      "auth": {
        "credential": true
      }
    },
    "headers": {
      "X-API-KEY": "api_key",
      "X-MISC-PASSWORD": "password"
    }
  }
}
```

## Error Handling

When credentials are missing or invalid, servers **SHOULD** respond with:

```json
{
  "jsonrpc": "2.0",
  "id": 1,
  "error": {
    "code": 32004,
    "message": "See required configuration",
    "data": {
      "requiredConfiguration": [
        {
          "name": "X-API-KEY",
          "description": "An API key must be provided to call this tool."
        }
      ]
    }
  }
}

> ```

Reviewer: See my comment/suggestion below about reporting OAuth errors—I think we should be uniform in this reporting and support presenting details from multiple auth schemes at once.

## Security Considerations

1. Clients **MUST**:
   - Securely obtain/store credentials
   - Only transmit credentials over secure channels

2. Servers **MUST**:

>    - Maintain secure storage of credentials if needed

Author: ```suggestion
   - Maintain secure storage of credentials if needed
   - Treat credential names as case-insensitive (like HTTP headers)
```

{{< /callout >}}

The Model Context Protocol (MCP) supports credential-based authentication for scenarios requiring API keys or similar secrets.

## Capabilities

Clients supporting credential authentication **MUST** declare it during initialization:

```json
{
  "capabilities": {
    "auth": {
      "credentials": true
    }
  }
}
```

Servers supporting credentials **MUST** include their capabilities:

```json
{
  "capabilities": {
    "auth": {
      "credentials": {
        "list": true
      }
    }
  }
}
```

## Protocol Messages

### Credential Requirements
Clients can list required credentials with an `auth/credentials/list` capability.

**Request:**
```json
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "auth/credentials/list",
}
```
**Response:**
```json
{
  "jsonrpc": "2.0",
  "id": 1,
  "result": {
    "credentials": [
      {
        "name": "API-KEY",
        "description": "An API key must be provided to call this tool."
      },
      {
        "name": "MISC-PASSWORD",
        "description": "A password must be provided to list this resource"
      }
    ]
  }
}
```
- Clients and servers **MUST** treat credential names as case-insensitive.

### Providing Credentials

> Clients **MUST** provide credentials through headers during initialization:

Reviewer: Mirroring the language in the OAuth doc:
```suggestion
During initialization, if the client and server both support the `credentials` capability, the client **SHOULD** provide credentials as parameters during initialization:
```

```json
{
  "jsonrpc": "2.0",
  "id": 1,
  "method": "initialize",
  "params": {
    "capabilities": {
      "auth": {

>         "credential": true

Reviewer: ```suggestion
        "credentials": true
```

      }
    },
    "auth": {
      "credentials": {
        "API-KEY": "api_key",
        "MISC-PASSWORD": "password"
      }
    }
  }
}
```

## Error Handling

When credentials are missing or invalid, servers **MUST** respond with at least an error code.

**Response:**

> ```json

Reviewer: ````suggestion
```typescript
````

(fixes the syntax highlighting when comments and things are included)

{
  "jsonrpc": "2.0",
  "id": 1,
  "error": {
    "code": -32001,
    "message": "Auth error, please see nested data.",
    "data": {
      "authRequest": {
        "credentials": {
          "error": "ASCII error code", // REQUIRED
          "errors": { // RECOMMENDED
            // Breakdown of error per-credential
          }

>           

Reviewer: I find this example a little hard to interpret. We don't have to be overly specific for now, since we can evolve this bit, but I'm curious what you have in mind here.

Author: Let's flesh this out in a followup then, since I don't have much on my mind here yet. Maybe something like this?

```
...
      "authRequest": {
        "credentials": {
          "error": "auth_error", 
          "errors": { // RECOMMENDED
            "API-KEY": "Missing",
            "MISC-PASSWORD": "Invalid"
          }
        }
...
```

  "params": {
    "capabilities": {
      "auth": {
        "credentials": true
      }
    },

>     "_meta": {

Reviewer: Sorry, I meant using the `_meta` field _outside_ of initialization only. 😅

Rationale:
1. It's an essential part of initialization semantics
2. It's an out of band addition to every other request type

I think just `params.auth` is fine during init.

Author: Hm, I think I would prefer consistency here. Providing it as part of `_meta` in initialization would be the same as making a request with headers. If the server finds valid credentials, it can present a larger suite of capabilities?

Reviewer: The thing is that the initialize request is basically _all_ headers (by analogy).
